File: src/main/java/com/thebluehats/server/game/world/ClearArrows.java
package com.thebluehats.server.game.world;

import org.bukkit.Bukkit;
import org.bukkit.entity.Arrow;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityShootBowEvent;
import org.bukkit.plugin.java.JavaPlugin;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.UUID;

public class ClearArrows implements Listener {
    private final HashMap<UUID, ArrayList<Arrow>> playerToArrows = new HashMap<>();

    private final JavaPlugin plugin;

    @Inject
    public ClearArrows(JavaPlugin plugin) {
        this.plugin = plugin;
    }

    @EventHandler
    public void onProjectileShoot(EntityShootBowEvent event) {
        Entity projectile = event.getProjectile();

        if (event.getProjectile() instanceof Arrow) {
            if (((Arrow) event.getProjectile()).getShooter() instanceof Player) {
                Player player = (Player) ((Arrow) event.getProjectile()).getShooter();

                if (!playerToArrows.containsKey(player.getUniqueId())) {
                    playerToArrows.put(player.getUniqueId(), new ArrayList<>());
                }

                ArrayList<Arrow> data = playerToArrows.get(player.getUniqueId());

                // TODO Server config
                int ARROW_MAXIMUM = 7;

                if (data.size() >= ARROW_MAXIMUM) {
                    data.get(0).remove();
                    data.remove(0);
                }

                data.add((Arrow) event.getProjectile());

                playerToArrows.put(player.getUniqueId(), data);
            }
        }

        Bukkit.getServer().getScheduler().scheduleSyncDelayedTask(plugin, () -> {
            if (projectile != null) {
                projectile.remove();
            }
        }, 100L);
    }
}

File: src/main/kotlin/com/thebluehats/server/game/world/ClearArrows.kt
package com.thebluehats.server.game.world

import org.bukkit.Bukkit
import org.bukkit.entity.Arrow
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityShootBowEvent
import org.bukkit.plugin.java.JavaPlugin
import java.util.*
import javax.inject.Inject

class ClearArrows @Inject constructor(private val plugin: JavaPlugin) : Listener {
    private val playerToArrows = HashMap<UUID, ArrayList<Arrow>>()
    @EventHandler
    fun onProjectileShoot(event: EntityShootBowEvent) {
        val projectile = event.projectile
        if (event.projectile is Arrow) {
            if ((event.projectile as Arrow).shooter is Player) {
                val player = (event.projectile as Arrow).shooter as Player
                if (!playerToArrows.containsKey(player.uniqueId)) {
                    playerToArrows[player.uniqueId] = ArrayList()
                }
                val data = playerToArrows[player.uniqueId]!!

                // TODO Server config
                val arrowMaximum = 7
                if (data.size >= arrowMaximum) {
                    data[0].remove()
                    data.removeAt(0)
                }
                data.add(event.projectile as Arrow)
                playerToArrows[player.uniqueId] = data
            }
        }
        Bukkit.getServer().scheduler.scheduleSyncDelayedTask(plugin, { projectile?.remove() }, 100L)
    }
}